### ISSUE TITLE
Cannot find module '$env/static/public' or its corresponding type declarations.ts(2307)

### ISSUE THREAD

**Issue by winston0410** [sveltejs/kit#8882]
### Describe the bug

The definition for '$env/static/public' cannot be found by Typescript in `*.svelte` file. It works in the runtime though

### Reproduction

https://github.com/winston0410/missing-env-module

### System Info

```Shell
System:
    OS: macOS 11.4
    CPU: (8) arm64 Apple M1
    Memory: 84.88 MB / 8.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.17.0 - ~/.nvm/versions/node/v16.17.0/bin/node
    Yarn: 1.22.17 - /opt/homebrew/bin/yarn
    npm: 8.15.0 - ~/.nvm/versions/node/v16.17.0/bin/npm
    Watchman: 2022.09.05.00 - /opt/homebrew/bin/watchman
  Browsers:
    Brave Browser: 109.1.47.186
    Safari: 14.1.1
  npmPackages:
    @sveltejs/adapter-auto: ^1.0.0 => 1.0.2 
    @sveltejs/kit: ^1.0.0 => 1.3.10 
    svelte: ^3.54.0 => 3.55.1 
    vite: ^4.0.0 => 4.1.1
```

### Severity

annoyance

**Comment by dummdidumm** [comment]
Did you run `vite dev` (for `npm run dev`) or `svelte-kit sync` prior to the error? The types need to be generated first - if I do this, I can't reproduce this.

**Comment by dummdidumm** [comment]
I'm using the latest version https://marketplace.visualstudio.com/items?itemName=svelte.svelte-vscode and no other Svelte-related extensions. I don't get this error inside Svelte files, once the types are generated.

**Comment by cesarnml** [comment]
<img width="1465" alt="Screenshot 2023-02-10 at 04 18 00" src="https://user-images.githubusercontent.com/26180499/217941318-d497faee-8e22-4246-b282-16b2cd08311d.png">

Same issue. This is only an issue with eslint. The module is imported properly, just give a false lint warning.

**Comment by winston0410** [comment]
For me I figured out the issue. Install the latest language tool and restarting it fixed the issue. I guess @cesarnml your issue is slightly different with mine? not sure

**Comment by Destaq** [comment]
Same problem here with a `.ts` file (in a Svelte project). Restarting, unfortunately, didn't fix anything.

**Comment by Destaq** [comment]
Here are a couple of workarounds:

1. Redefine the module in your `src` folder:
```ts
// src/types.d.ts
declare module "$env/static/public" {
     export const PUBLIC_KEY: <type>;
}
```

2. Explicitly include the generated types file in your `tsconfig.json`:
```json5
{
    "include": ["src/**/*", "src/node_modules", ".svelte-kit/ambient.d.ts"], // see last element
}
```

Not sure why it doesn't work out of the box though.

**Comment by Rich-Harris** [comment]
Can anyone reproduce this in a fresh project? I can't

**Comment by DanielRios549** [comment]
Why `env.publicPrefix` is not `_VITE`by default since this is the default from Vite?

**Comment by homerjam** [comment]
![Screenshot 2023-03-16 at 17 08 12](https://user-images.githubusercontent.com/1055769/225702519-c731dfff-852a-41d0-8d14-82eef097bab8.jpg)

I have a possibly related issue

**Comment by dummdidumm** [comment]
Which version of SvelteKit are you on? The error seems unrelated and the cause for it should be fixed in the latest version.

**Comment by benwoodward** [comment]
adding `".svelte-kit/ambient.d.ts"` to the list of includes in `.tsconfig.json` fixed this for me.

**Comment by JJBoe** [comment]
> Can anyone reproduce this in a fresh project? I can't

Unfortunately, yes.
Below a screenshot of a "fresh" [DENO fresh](https://fresh.deno.dev/) installation, without any alterations.
`deno run -A -r https://fresh.deno.dev my-project`

Have you come across any information in the meantime that might solve this problem?


![2023-04-30_18h57_36](https://user-images.githubusercontent.com/56443470/235366055-9d773c1a-5f28-44d6-8685-ce6f254d405e.png)
![2023-04-30_18h58_58](https://user-images.githubusercontent.com/56443470/235366102-33e52a5f-4424-4c05-a0b3-a9a3f9a85f0c.png)

**Comment by skylar** [comment]
I also ran into these issues today when updating to the latest Svelte/kit and Vite libraries.

**Comment by amyjko** [comment]
Likely due to the new 0.32 breaking change, which throws errors on unresolvable modules:

https://github.com/vitest-dev/vitest/releases/tag/v0.32.0

It looks like vitest was masking unresolved modules, and now it's not.

**Comment by phobetron** [comment]
Downgrading vitest to `v0.31.4` fixed a similar problem for my project.

**Comment by DanielRios549** [comment]
I fixed this problem by adding both `.svelte-kit/ambient.d.ts` and `.svelte-kit/types/**/$types.d.ts` in the `include` section of `tsconfig.json`.

Make sure `.svelte-kit` folder is not in the `exclude` section, otherwise it will remove both entries above.

**Comment by MiguelArmendariz** [comment]
I also met this issue, I tried the work arounds with no luck. 
The env vars work, so I think I'll continue working as it is.

```
"devDependencies": {
		"@playwright/test": "^1.28.1",
		"@sveltejs/adapter-auto": "^2.0.0",
		"@sveltejs/kit": "^1.24.0",
		"@typescript-eslint/eslint-plugin": "^6.5.0",
		"@typescript-eslint/parser": "^6.5.0",
		"autoprefixer": "^10.4.15",
		"eslint": "^8.48.0",
		"eslint-config-prettier": "^9.0.0",
		"eslint-plugin-svelte": "^2.30.0",
		"postcss": "^8.4.29",
		"prettier": "^3.0.3",
		"prettier-plugin-svelte": "^3.0.3",
		"svelte": "^4.0.5",
		"svelte-check": "^3.5.1",
		"tailwindcss": "^3.3.3",
		"tslib": "^2.4.1",
		"typescript": "^5.2.2",
		"vite": "^4.4.2",
		"vitest": "^0.34.3"
	},
```

**Comment by 524c** [comment]
> ```js
> "src/node_modules", ".svelte-kit/ambient.d.ts"
> ```

this fixes the TS warning for me.

**Comment by boian-ivanov** [comment]
This seems to be caused only if you override the `tsconfig`'s `include` path, in essence removing the `ambient.d.ts`, which you would extend previously via the generated `./.svelte-kit/tsconfig.json`. Had it as well and just removing the `include` worked great.

**Comment by xxxhussein** [comment]
> Can anyone reproduce this in a fresh project? I can't

i got it when i used "plugin:@typescript-eslint/strict-type-checked"

**Comment by xxxhussein** [comment]
and also i got an insane amount of errors from using this config, mainly the issue above and all variables that are correctly typed by sveltekit are `any` according to eslint

**Comment by dummdidumm** [comment]
Everyone who also has the problem, please always tell if you have the `include` option in your `tsconfig.json` and if yes what it contains - I suspect that all these errors come from adding `include` to the `tsconfig.json`. It should not be necessary to do this, and if you do, it will override any of the `include` options that are prepared for you in the generated `tsconfig.json` that's within the `.svelte-kit` folder.

**Comment by 524c** [comment]
> Everyone who also has the problem, please always tell if you have the `include` option in your `tsconfig.json` and if yes what it contains - I suspect that all these errors come from adding `include` to the `tsconfig.json`. It should not be necessary to do this, and if you do, it will override any of the `include` options that are prepared for you in the generated `tsconfig.json` that's within the `.svelte-kit` folder.

Mine was like this:
```
  "include": [
    "./**/*.ts",
    "./**/*.js",
  ]

```
And it had resolved after adjusting to:

"./.svelte-kit/tsconfig.json"
```
  "include": [
    "./**/*.ts",
    "./**/*.js",
    ".svelte-kit/ambient.d.ts",
  ]
```

**Comment by 524c** [comment]
I just confirmed that when I completely remove my include the problem goes away.

**Comment by ZerdoX-x** [comment]
@dummdidumm 

My case:
1) I was trying to deal with https://github.com/sveltejs/eslint-plugin-svelte/issues/432 by adding and configuring https://github.com/ota-meshi/typescript-eslint-parser-for-extra-files
2) It required me to set (see project's readme)
```json
"include": [
    "**/*.svelte",
    "**/*.ts",
    "**/*.tsx"
]
```

Can confirm that adding `".svelte-kit/ambient.d.ts"` to `"includes"` resolves the issue.
Should we just add `".svelte-kit/ambient.d.ts"` to `tsconfig.includes` by default in all svelte project templates? Because sometimes users need to define their own "includes" for the project, but tsconfig just redefines the whole field. Never liked how tsconfig works.. It is a nightmare


For me modifying starter templates and maybe [this page](https://svelte.dev/docs/typescript) of documentation (could not find similar page for sveltekit except [this one](https://kit.svelte.dev/docs/types)) looks like a solution to satisfy the issue and DX.
And I hope we won't leave it as "works as intended" due to typescript's tsconfig nature :D

**Comment by juwul** [comment]
@ZerdoX-x 
> Can confirm that adding ".svelte-kit/ambient.d.ts" to "includes" resolves the issue.

Same experience for me in a completely fresh project.

```	
"devDependencies": {
		"@angular/cli": "^17.1.0",
		"@sveltejs/adapter-auto": "^3.0.0",
		"@sveltejs/kit": "^2.0.0",
		"@sveltejs/vite-plugin-svelte": "^3.0.0",
		"svelte": "^4.2.7",
		"svelte-check": "^3.6.0",
		"tslib": "^2.4.1",
		"typescript": "^5.0.0",
		"vite": "^5.0.3"
	},
```

**Comment by pboling** [comment]
Same experience for me in a brand new SvelteKit 2 / Svelte 5 project.  I have not modified any `tsconfig.json`.  I installed `svelte-check`, ran it (no issues found), installed a few Svelte-related JetBrains WebStorm IDE plugins, restarted my IDE, switched branches working on other stuff a few times, and eventually came back, re-did `pnpm install`, and noticed I don't have the warning anymore!

**Comment by Davidiusdadi** [comment]
Had the same issue - and **solved it like this**:

For some reason I had added `".svelte-kit"` to my repo root tsconfig **exclude** list. :man_facepalming: 

By default `"extends": "./.svelte-kit/tsconfig.json",` in the tsconfig should handle things:
That means that `yarn tsc --showConfig` will list files `.svelte-kit/*` like `.svelte-kit/ambient.d.ts` within the `files` entry!

If not, the files have:
- not yet been generated (as noted by [dummdidumm](https://github.com/dummdidumm)  )
- are not included because you override `include` shadowing the `./.svelte-kit/tsconfig.json` include.
- are _maybe_ included but are excluded via `"exclude": [".svelte-kit"]`

**Comment by arkmech** [comment]
Yeah this doesn't work at all with new project:

1. Root `tsconfig.json`, which extends `"./.svelte-kit/tsconfig.json"`
2. `.svelte-kit` tsconfig.json includes `ambient.d.ts`
3. `ambient.d.ts` declares module `'$env/static/private'` that includes my env variables in my `.env`.

I get `Cannot find module '$env/static/private' or its corresponding type declarations.`

```
	"devDependencies": {
		"@biomejs/biome": "^1.6.0",
		"@playwright/test": "^1.28.1",
		"@sveltejs/adapter-auto": "^3.0.0",
		"@sveltejs/kit": "^2.5.3",
		"@sveltejs/vite-plugin-svelte": "^3.0.0",
		"@tailwindcss/vite": "4.0.0-alpha.7",
		"drizzle-kit": "^0.20.14",
		"svelte": "^5.0.0-next.74",
		"svelte-check": "^3.6.0",
		"tailwindcss": "4.0.0-alpha.7",
		"tslib": "^2.4.1",
		"typescript": "^5.0.0",
		"vite": "^5.0.3",
		"vitest": "^1.2.0"
	}
```

**Comment by eltigerchino** [comment]
> EDIT: This is error is coming from my `drizzle.config.ts` which is located at root of the project. Maybe that's why? Is `$env/static/private` only available within `src`?

Yes, your drizzle config typescript file needs to be added to the `include` array in tsconfig. But you don't want to override the include array that sveltekit generates, so you need to copy over the ones inside `./.svelte-kit/tsconfig.json` and then add an additional entry.

**Comment by Bewinxed** [comment]
> Had the same issue - and **solved it like this**:
> 
> For some reason I had added `".svelte-kit"` to my repo root tsconfig **exclude** list. 🤦‍♂️
> 
> By default `"extends": "./.svelte-kit/tsconfig.json",` in the tsconfig should handle things: That means that `yarn tsc --showConfig` will list files `.svelte-kit/*` like `.svelte-kit/ambient.d.ts` within the `files` entry!
> 
> If not, the files may have:
> 
> * not yet been generated (as noted by [dummdidumm](https://github.com/dummdidumm)  )
> * not been included because you override `include` shadowing the `./.svelte-kit/tsconfig.json` include.
> * _maybe_ been included but are excluded via `"exclude": [".svelte-kit"]`

Thanks for this, Copilot added it in 🤦‍♂️

**Comment by mhamri** [comment]
As far as I understand, the include option won't be merged; instead, it will be overridden. This means that if you have an include in your .sveltekit. tsconfig.json, it will be overridden by the one in root. 

so in my case the culprit was having `include` and `exclude` in the root tsconfig.json

https://www.typescriptlang.org/tsconfig/#extends

**Comment by AlexRMU** [comment]
I solved the problem when I moved `.svelte-kit` from `node_modules` back to the root.
TS sometimes has a problem with importing generated types from `node_modules`, it also occurred when working with @prisma.

**Comment by Harold-Anderson** [comment]
I solved the problem by removing both `includes `and `excludes `from `tsconfig.json` 

Thanks to @dummdidumm  https://github.com/sveltejs/kit/issues/8882#issuecomment-1848443983

**Comment by joelhickok** [comment]
Problem still exists.  I never had it before, but after upgrading my project to Svelte v.5, as well as several other related modules, this came up.
Resolved by adding `".svelte-kit/ambient.d.ts"` to the include property in `tsconfig.json`.

**Comment by eltigerchino** [comment]
> Problem still exists. I never had it before, but after upgrading my project to Svelte v.5, as well as several other related modules, this came up. Resolved by adding `".svelte-kit/ambient.d.ts"` to the include property in `tsconfig.json`.

Can you create a minimal reproduction? Did you generate the tsconfig beforehand by running `pnpm dev` or `pnpm check`?

**Comment by MK10UNoY** [comment]
For my export related issue where no exported members from .env were detected, i tried fixing it by reload but nothing worked
but finally saw this section in .svelte-kit\ambient.d.ts and added this line 

declare module '$env/static/public' {
	export const PUBLIC_SUPABASE_URL: string;
	export const PUBLIC_SUPABASE_ANON_KEY: string;
}
that seemed to fix the issue.

**Comment by joelhickok** [comment]
> > Problem still exists. I never had it before, but after upgrading my project to Svelte v.5, as well as several other related modules, this came up. Resolved by adding `".svelte-kit/ambient.d.ts"` to the include property in `tsconfig.json`.
> 
> Can you create a minimal reproduction? Did you generate the tsconfig beforehand by running `npm run dev` or `npm run check`?

`tsconfig.json` existed already from the original svelte app a long time ago.  This came up when upgrading to Svelte 5, but I also upgraded some other packages at the same time.  Not sure if this was happening only when upgrading an existing project to Svelte 5.

**Comment by longrunningprocess** [comment]
https://github.com/sveltejs/kit/issues/8882#issuecomment-2420772631 worked for me

**Comment by tawsbob** [comment]
just add your env vars to the .env guys

**Comment by joelhickok** [comment]
> just add your env vars to the .env guys

Yeah, but that's not the point.  The developers designed a system and it wasn't working as expected.  If people start to use the features you develop based on the documentation, and it breaks, then yes, you can go back to the tried and true method.  Before the import system  using $env/static/public was created, I used .env files.

**Comment by teemingc** [comment]
> > just add your env vars to the .env guys
> 
> Yeah, but that's not the point. The developers designed a system and it wasn't working as expected. If people start to use the features you develop based on the documentation, and it breaks, then yes, you can go back to the tried and true method. Before the import system using $env/static/public was created, I used .env files.

Can you share a minimal reproduction?

**Comment by mikayil-dev** [comment]
> Problem still exists. I never had it before, but after upgrading my project to Svelte v.5, as well as several other related modules, this came up. Resolved by adding `".svelte-kit/ambient.d.ts"` to the include property in `tsconfig.json`.

This fixed it for me

**Comment by PaperPrototype** [comment]
I just had to move my hooks.server.ts into the src/ folder, I was in the root folder. I also needed to run `npm run dev`

**Comment by teemingc** [comment]
Closing this as stale. If anyone comes across this issue, you've probably overridden the `tsconfig.json` include/exclude. You should instead modify the setting via https://svelte.dev/docs/kit/configuration#typescript
